warehouse/query-core/src/main/java/datawave/next/DocIdIteratorVisitor.java

```diff
@@ -0,0 +1,521 @@
+
+import datawave.next.stats.DocumentIteratorStats;
+import datawave.query.jexl.JexlASTHelper;
+import datawave.query.jexl.nodes.QueryPropertyMarker;
+import datawave.query.jexl.visitors.BaseVisitor;
+import datawave.query.jexl.visitors.JexlStringBuildingVisitor;
+
+/**
+ * A visitor that scans the field index and returns all document ids that match a given query.
+ * <p>
+ * Operators that are supported
+ * <ul>
+ * <li>equality</li>
+ * <li>regex</li>
+ * <li>range</li>
+ * <li>list marker</li>
+ * <li>negations that are part of an intersection</li>
+ * </ul>
+ * <p>
+ * Operators that are NOT supported
+ * <ul>
+ * <li>negated regex</li>
+ * <li>negated equality</li>
+ * <li>functions</li>
+ * <li>term markers</li>
+ * </ul>
+ */
+public class DocIdIteratorVisitor extends BaseVisitor {
+
+    private static final Logger log = LoggerFactory.getLogger(DocIdIteratorVisitor.class);
+
+    private final Range range;
+    private final SortedKeyValueIterator<Key,Value> source;
+    private final Set<String> datatypeFilter;
+    private final LongRange timeFilter;
+    private final Set<String> indexedFields;
+
+    private final String row;
+    private boolean isDocRange = false;
+
+    private long maxScanTimeMillis = 15_000L;
+    private long resultInterval = 500;
+
+    private final DocumentIteratorStats stats = new DocumentIteratorStats();
+
+    /**
+     *
+     * @param script
+     *            the query tree
+     * @param range
+     *            the range
+     * @param source
+     *            the source iterator
+     * @param datatypeFilter
+     *            the datatype filter
+     * @param timeFilter
+     *            the time filter
+     * @return the set of document ids that satisfy the query
+     */
+    public static Set<Key> getDocIds(ASTJexlScript script, Range range, SortedKeyValueIterator<Key,Value> source, Set<String> datatypeFilter,
+                    LongRange timeFilter, Set<String> indexedFields) {
+        DocIdIteratorVisitor visitor = new DocIdIteratorVisitor(source, range, datatypeFilter, timeFilter, indexedFields);
+        Object o = script.jjtAccept(visitor, null);
+        if (o instanceof ScanResult) {
+            return ((ScanResult) o).getResults();
+        }
+        return Collections.emptySet();
+    }
+
+    public Set<Key> getDocIds(ASTJexlScript script) {
+        Object o = script.jjtAccept(this, null);
+        if (o instanceof ScanResult) {
+            return ((ScanResult) o).getResults();
+        }
+        return Collections.emptySet();
+    }
+
+    protected DocIdIteratorVisitor(SortedKeyValueIterator<Key,Value> source, Range range, Set<String> datatypeFilter, LongRange timeFilter,
+                    Set<String> indexedFields) {
+        this.source = source;
+        this.range = range;
+        this.datatypeFilter = datatypeFilter;
+        this.timeFilter = timeFilter;
+        this.indexedFields = indexedFields;
+
+        this.row = range.getStartKey().getRow().toString();
```

Comment:
check precondition that start key in the range is not null?

Reply:
https://github.com/NationalSecurityAgency/datawave/issues/2916